src/download.rs

```diff
@@ -391,10 +391,146 @@ struct BlockContext {
+    fn new_target<P: AsRef<Path>>(path: P) -> Result<Box<dyn SnapshotWriteTarget>> {
+        let path = path.as_ref();
+        Ok(Box::new(BlockDeviceTarget { path: path.into() }))
+    }
+
+    async fn is_valid<P: AsRef<Path>>(path: P) -> Result<bool> {
+        let path = path.as_ref();
+        if !path.exists() {
+            return Ok(false);
+        }
+
+        let file_meta = fs::metadata(path)
+            .await
+            .context(error::ReadFileMetadata { path })?;
+
+        if file_meta.file_type().is_block_device() {
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+}
+
+#[async_trait]
+impl SnapshotWriteTarget for BlockDeviceTarget {
+    // ensures existing size >= length, but otherwise leaves untouched
+    async fn grow(&mut self, length: i64) -> Result<()> {
+        let path = self.path.as_path();
+        let block_device_size =
+            get_block_device_size(path).context(error::GetBlockDeviceSize { path })?;
+
+        // Make sure the block device is big enough to hold the snapshot
+        ensure!(
+            block_device_size >= length,
+            error::BlockDeviceTooSmall {
+                block_device_size: block_device_size / GIBIBYTE,
+                needed: length / GIBIBYTE,
+            }
+        );
+
+        Ok(())
+    }
+
+    // returns the file path to which blocks must be written
+    fn write_path(&self) -> Result<&Path> {
+        Ok(self.path.as_path())
+    }
+
+    // no-op
+    fn finalize(&mut self) -> Result<()> {
+        Ok(())
+    }
+}
+
+/// Implements file operations for filesystem files.
+struct FileTarget {
+    path: PathBuf,
+    temp_file: Option<NamedTempFile>,
+}
+
+impl FileTarget {
+    fn new_target<P: AsRef<Path>>(path: P) -> Result<Box<dyn SnapshotWriteTarget>> {
+        let path = path.as_ref();
+        Ok(Box::new(FileTarget {
+            path: path.into(),
+            temp_file: None,
+        }))
+    }
+}
+
+#[async_trait]
+impl SnapshotWriteTarget for FileTarget {
+    // truncate file to desired size
+    async fn grow(&mut self, length: i64) -> Result<()> {
+        let path = self.path.as_path();
+
+        // Create a temporary file and extend it to the required size.
+        let target_dir = path
+            .parent()
+            .context(error::ValidateParentDirectory { path })?;
+
+        let temp_file = NamedTempFile::new_in(target_dir)
+            .context(error::CreateTempFile { path: target_dir })?;
+
+        let temp_file_len = length;
+        let temp_file_len = u64::try_from(temp_file_len).with_context(|| error::ConvertNumber {
+            what: "temp file length",
+            number: temp_file_len.to_string(),
+            target: "u64",
+        })?;
+
+        temp_file
+            .as_file()
+            .set_len(temp_file_len)
+            .context(error::ExtendTempFile {
+                path: temp_file.as_ref(),
+            })?;
+
+        self.temp_file.replace(temp_file);
```

Comment:
Should it error if `self.temp_file.is_some()`? Can `grow` be called more than once before `finalize` without a problem?

Edit: oh I guess you're just moving this code...